pkg/controller/nodepodprobe/node_pod_probe_controller.go

```diff
@@ -0,0 +1,362 @@
+	// map[pod.uid]=Pod
+	matchedPods := map[string]*corev1.Pod{}
+	for _, obj := range npp.Spec.PodProbes {
+		pod := &corev1.Pod{}
+		err := r.Get(context.TODO(), client.ObjectKey{Namespace: obj.Namespace, Name: obj.Name}, pod)
+		if err != nil && !errors.IsNotFound(err) {
+			klog.Errorf("NodePodProbe get pod(%s/%s) failed: %s", obj.Namespace, obj.Name, err.Error())
+			return nil, err
+		}
+		if errors.IsNotFound(err) || !kubecontroller.IsPodActive(pod) || string(pod.UID) != obj.UID {
+			continue
+		}
+		matchedPods[string(pod.UID)] = pod
+	}
+
+	newSpec := appsv1alpha1.NodePodProbeSpec{}
+	for i := range npp.Spec.PodProbes {
+		obj := npp.Spec.PodProbes[i]
+		if _, ok := matchedPods[obj.UID]; ok {
+			newSpec.PodProbes = append(newSpec.PodProbes, obj)
+		}
+	}
+	if reflect.DeepEqual(newSpec, npp.Spec) {
+		return matchedPods, nil
+	}
+
+	nppClone := &appsv1alpha1.NodePodProbe{}
+	err := retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		if err := r.Client.Get(context.TODO(), types.NamespacedName{Name: npp.Name}, nppClone); err != nil {
+			klog.Errorf("error getting updated npp %s from client", npp.Name)
+		}
+		if reflect.DeepEqual(newSpec, nppClone.Spec) {
+			return nil
+		}
+		nppClone.Spec = newSpec
+		return r.Client.Update(context.TODO(), nppClone)
+	})
+	if err != nil {
+		klog.Errorf("NodePodProbe update NodePodProbe(%s) failed:%s", npp.Name, err.Error())
+		return nil, err
+	}
+	klog.V(3).Infof("NodePodProbe update NodePodProbe(%s) from(%s) -> to(%s) success", npp.Name, util.DumpJSON(npp.Spec), util.DumpJSON(newSpec))
+	return matchedPods, nil
+}
+
+func (r *ReconcileNodePodProbe) updatePodProbeStatus(pod *corev1.Pod, status appsv1alpha1.PodProbeStatus) error {
+	// map[probe.name]->probeState
+	currentConditions := make(map[string]appsv1alpha1.ProbeState)
+	for _, condition := range pod.Status.Conditions {
+		currentConditions[string(condition.Type)] = appsv1alpha1.ProbeState(condition.Status)
+	}
+	type metadata struct {
+		Labels      map[string]interface{} `json:"labels,omitempty"`
+		Annotations map[string]interface{} `json:"annotations,omitempty"`
+	}
+	// patch labels or annotations in pod
+	probeMetadata := metadata{
+		Labels:      map[string]interface{}{},
+		Annotations: map[string]interface{}{},
+	}
+	// pod status condition record probe result
+	var probeConditions []corev1.PodCondition
+	var err error
+	for i := range status.ProbeStates {
+		probeState := status.ProbeStates[i]
+		// ignore the probe state
+		if probeState.State == "" || probeState.State == currentConditions[probeState.Name] {
+			continue
+		}
+
+		var conStatus corev1.ConditionStatus
+		if probeState.State == appsv1alpha1.ProbeSucceeded {
+			conStatus = corev1.ConditionTrue
+		} else {
+			conStatus = corev1.ConditionFalse
+		}
+		ppmName, probeName := strings.Split(probeState.Name, "#")[0], strings.Split(probeState.Name, "#")[1]
```

Comment:
since we can get pod probe marker name from NodePodProbe.probes[*].podProbeMarkerName, why we have to encode the podProbeMarkerName in the probe name ?